galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,478 @@
+    if not isinstance(world_pos, galsim.CelestialCoord):
+        raise TypeError("Position on the sky must be given as a galsim.CelestialCoord!")
+
+    # Compute position angle of FPA f2 axis, where positive corresponds to the angle east of North.
+    if PA_is_FPA:
+        pa_fpa = PA
+        pa_obsy = PA - 90.*galsim.degrees - theta_fpa
+    else:
+        pa_obsy = PA
+        pa_fpa = PA + 90.*galsim.degrees + theta_fpa
+    cos_pa = np.cos(pa_fpa.rad())
+    sin_pa = np.sin(pa_fpa.rad())
+
+    # Figure out tangent-plane positions for FPA center:
+    xc_fpa_tp, yc_fpa_tp = _det_to_tangplane_positions(xc_fpa, yc_fpa)
+
+    # Note, this routine reads in the coeffs.  We don't use them until later, but read them in for
+    # all SCAs at once.
+    a_sip, b_sip = _parse_sip_file(sip_filename)
+
+    # Loop over SCAs:
+    wcs_dict = {}
+    for i_sca in SCAs:
+        # Set up the header.
+        header = galsim.FitsHeader()
+        # Populate some necessary variables in the FITS header that are always the same, regardless of
+        # input and SCA number.
+        _populate_required_fields(header)
+
+        # And populate some things that just depend on the overall locations or other input, not on
+        # the SCA.
+        header['RA_TARG'] = (world_pos.ra / galsim.degrees,
+                             "right ascension of the target (deg) (J2000)")
+        header['DEC_TARG'] = (world_pos.dec / galsim.degrees,
+                              "declination of the target (deg) (J2000)")
+        header['PA_OBSY'] = (pa_obsy / galsim.degrees, "position angle of observatory Y axis (deg)")
+        header['PA_FPA'] = (pa_fpa / galsim.degrees, "position angle of FPA Y axis (deg)")
+
+        # Finally do all the SCA-specific stuff.
+        header['SCA_NUM'] = (i_sca, "SCA number (1 - 18)")
+
+        # Set the position of center of this SCA in focal plane angular coordinates.
+        sca_xc_fpa = np.arctan(sca_xc_mm[i_sca]/focal_length)*galsim.radians
+        sca_yc_fpa = np.arctan(sca_yc_mm[i_sca]/focal_length)*galsim.radians
+
+        # Figure out tangent plane positions after distortion, and subtract off those for FPA center
+        # (calculated in header).
+        sca_xc_tp, sca_yc_tp = _det_to_tangplane_positions(sca_xc_fpa, sca_yc_fpa)
+        # These define the tangent plane (X, Y) distance of the center of this SCA from the center
+        # of the overall FPA.
+        sca_xc_tp_f = sca_xc_tp - xc_fpa_tp
+        sca_yc_tp_f = sca_yc_tp - yc_fpa_tp
+
+        # Leave phi_p at 180 (0 if dec_targ==-90), so that tangent plane axes remain oriented along
+        # celestial coordinates. In other words, phi_p is the angle of the +Y axis in the tangent
+        # plane, which is of course pi if we're measuring these phi angles clockwise from the -Y
+        # axis.
+        if world_pos.dec / galsim.degrees > -90.:
+            phi_p = np.pi*galsim.radians
+        else:
+            phi_p = 0.*galsim.radians
```

Comment:
I have to be honest - I agree with you that it�s a bit odd.  The LONPOLE keyword is in the WCS routines that come from the WFIRST project office.  It does figure into their calculations, but not ours.  If we want to have our headers look like theirs (once I get the extra keywords going to file) then we should include it, but that�s really the only reason to do so.  I am adding a comment explaining this.